test/test_cpp_api_parity.py

```diff
@@ -0,0 +1,456 @@
+import copy
+import unittest
+import warnings
+import inspect
+
+import torch
+from torch._six import PY2
+import common_utils as common
+import common_nn
+from common_cuda import TEST_CUDA
+import torch.utils.cpp_extension
+from cpp_api_parity import sample_module, torch_nn_modules, TorchNNTestParams, CppArg
+
+
+torch_nn_has_parity = set([
+    'SampleModule',
+])
+
+TORCH_NN_MODULE_COMMON_TEST_HARNESS = """\n
+#include <torch/script.h>
+
+const char * const parity_test_error_msg = "Parity test failed";
+
+bool check_tensor_equality(const torch::Tensor& tensor1, const torch::Tensor& tensor2) {
+  return tensor1.sizes().vec() == tensor2.sizes().vec() && \
+    tensor1.device() == tensor2.device() && \
+    tensor1.dtype() == tensor2.dtype() && \
+    tensor1.allclose(tensor2);
+}
+
+bool check_ivalue_equality(const c10::IValue& ivalue1, const c10::IValue& ivalue2) {
+  if (ivalue1.tagKind() != ivalue2.tagKind()) {
+    AT_ERROR("Value type mismatch: ", "ivalue1: ", ivalue1.tagKind(), ", ivalue2: ", ivalue2.tagKind());
+  }
+  if (ivalue1.isInt()) {
+    return ivalue1.toInt() == ivalue2.toInt();
+  } else if (ivalue1.isDouble()) {
+    return ivalue1.toDouble() == ivalue2.toDouble();
+  } else if (ivalue1.isBool()) {
+    return ivalue1.toBool() == ivalue2.toBool();
+  } else if (ivalue1.isString()) {
+    return ivalue1.toString() == ivalue2.toString();
+  } else if (ivalue1.isTensor()) {
+    return check_tensor_equality(ivalue1.toTensor(), ivalue2.toTensor());
+  } else {
+    AT_ERROR("Unsupported value type: ", ivalue1.tagKind());
+  }
+}
+"""
+
+CHECK_MODULE_PARAM_EQUALITY = Template("""\
+TORCH_CHECK(
+  check_tensor_equality(${script_module_prefix}.get_parameter("${param_name}"), ${cpp_module_prefix}->${param_name}),
+  parity_test_error_msg);
+""")
+
+CHECK_MODULE_BUFFER_EQUALITY = Template("""\
+TORCH_CHECK(
+  check_tensor_equality(${script_module_prefix}.get_buffer("${buffer_name}"), ${cpp_module_prefix}->${buffer_name}),
+  parity_test_error_msg);
+""")
+
+CHECK_MODULE_ATTR_EQUALITY = Template("""\
+TORCH_CHECK(
+  check_ivalue_equality(
+    ${script_module_prefix}.get_attribute("${attr_name}"), c10::IValue(${cpp_module_prefix}->${attr_name})),
+  parity_test_error_msg);
+""")
+
+TORCH_NN_MODULE_TEST_CTOR_ARGS = Template("""\n
+void ${module_name}_test_ctor_args() {
+  ${module_qualified_name} m_init_by_cpp(${module_option});
+}
+""")
+
+TORCH_NN_MODULE_TEST_INIT = Template("""\n
+void ${module_variant_name}_test_init(
+    const std::string& saved_module_path,
+    const std::string& device) {
+  torch::jit::script::Module m_init_by_python = torch::jit::load(saved_module_path);
+
+  torch::manual_seed(2);
+  ${module_qualified_name} m_init_by_cpp${cpp_constructor_args};
+  m_init_by_cpp->to(device);
+
+  ${extra_stmts}
+}
+""")
+
+TORCH_NN_MODULE_TEST_FORWARD = Template("""\n
+void ${module_variant_name}_test_forward(
+    const std::string& saved_module_path,
+    const std::string& device,
+    torch::Tensor python_output,
```

Comment:
I'm not sure if all modules return tensors (LSTM, GRU, and RNN don't), but we can cross that bridge in a followup PR when someone adds tests for those.

Reply:
Will do when someone adds tests for those modules.